New feature to handle circular references individually for each item upon collection serialization

jobosk opened the pull request:
As I've mentioned in issue #2791 I propose the addition of a new feature to handle circular references for each item individually upon collections' serialization, in order to retrieve all necessary information in each iteration, but still avoiding recursive serialization due to circular references.

cowtowncoder commented:
I am not sure I still understand the use case, but some quick notes:

* this can not really go in 2.9 branch (2.12 is the only one that can add new features)
* unit tests would be needed to show expected usage

I think I could also accept new method for clearing references (so first part of changes here), to be called manually by custom serializers.

Also: come to think of that... I really do not get why resetting should commonly be done for each entry (I could maybe kind of see it on per-collection, i.e. before serializing elements), but... to avoid keeping references across entries? I do not understand the use case at all.

kupci commented:
@jobosk It would be helpful if you had some unit tests to show the use case you are trying to support. Maybe there's another way to solve this?

jobosk commented:
I've added a unit test trying to show what I consider "expected behaviour" and currently can only achieve with the feature I implemented.
I've extended such behaviours to all collection serializers.

jobosk commented:
@kupci @tatu-at-salesforce Did you have time to review the last PR? it includes a test showing what I'm trying to achieve, and also, it would be an optional feature, not default behaviour.
Please, let me know whether if I explained myself, wether or not you agree. Thanks!

cowtowncoder commented:
I had a quick look and no, I don't think I'd accept this. Too specific for an on/off feature. Also noticed that this clears the whole per context references for every element of any collection. It does not establish any other scoping, just removes all mappings for everything for every element.

I don't know how it would be done in a better way but I don't really want to add such implementation.

jobosk commented:
That's a shame.
What you describe is exactly what I intended, "remembering" references solely within the scope (not Jackson-scope, English language meaning of "scope") of each item from the collection (which means "forgetting" them each iteration), so that the circular dependency handling in one iteration doesn't affect the serialisation of the rest.
I don't know a better way to do it either, I was constraint due to the current strategy, but it does the trick for me.

I still don't understand why it has been chosen as default behaviour to make the serialisation of collections non deterministic (the same item will not be serialised the same way every time, but will depend on the rest of the items, in particular those ahead from it and the order in which they appear).
